### DISCUSSION TITLE
[ENH] Is there any interest in adding an error handling option within `EnsembleForecaster`?

### DISCUSSION THREAD

**Discussion by ninedigits** [alan-turing-institute/sktime#5729] [Ideas]
In the `EnsembleForecaster`, when an exception occurs, the code halts. This is often intentional, allowing for adjustments in preprocessing or model substitution. However, it becomes challenging in tasks where I forecast hundreds or thousands of time series. Specifically, tailoring a model to each individual time series is often impractical. Using the same `EnsembleForecaster` (fit to each time series) not only reduces the risk of overfitting but also simplifies maintenance.

Previously, I've handled this with a try-except block::

```python3
forecaster = EnsembleForecaster([
    ("ets", AutoETS(auto=True, n_jobs=1, additive_only=True)),
    ("naive_mean", NaiveForecaster(strategy="last"))
  ]
)

forecaster.fit()
try:
    y_pred = forecaster.predict():
except:
    # use different ensemble forecaster combo ...
```
As a potential enhancement, I propose adding an error handling parameter to `EnsembleForecaster`, such as `errors="ignore"`, `errors="raise"`, or `errors="warn"`. This would enable the forecaster to continue producing forecasts even if one component fails, unless all or a specified number of models fail. Example:

```python3

forecaster = EnsembleForecaster([
    ("ets", AutoETS(auto=True, n_jobs=1, additive_only=True)),
    ("naive_mean", NaiveForecaster(strategy="last"))
  ], errors="ignore" #or "raise" or "warn"
)
```

Would this feature be of interest to others? Happy to work on a PR to develop it if there's an interest in it.

**Comment by fkiraly** [comment]
Very nice idea!

I see two options that could be of use:

* adding the error handling parameter as described, need to be careful with means and the like (denominator)
* decomposing things into multiple compoets:
    * to `DummyForecaster`, add a "constant" `strategy`, with a default value of `np.nan` (?) or a settable value `constant_value` if that strategy is used
    * in addition, we ensure that `EnsembleForecaster` is fine with `np.nan`-s
    * we add a `FallbackForecaster` that implements generic error handling - one passes a list of forecasters, and if a forecaster fails, the next in the list is attempted (needs parameters like "should I fit all at the start", "does fallback happen in `fit` or in `predict`" etc). This would inherit from `BaseMetaEstimator`.
    * then the proposed error handling can be built by falling back on the `nan` constant dummy - but also many many more variants of error handling logic become available to the user!

I would suggest though to start with the first option, it is less general and programmatic, but the 80 in the 80/20 here imo.

**Reply by ninedigits** [reply]
I agree that option 1 provides a quick and straightforward implementation, focusing on immediate needs. However, option 2 offers a more comprehensive and flexible solution, potentially benefiting other modules as well, but it's a bigger lift with more requirements and testing needs.

I really like the idea of a `FallbackForecaster`. I'm imagining its use might look something like this:

```python3
forecaster = FallbackForecaster(
    forecasters=[
        # First try: EnsembleForecaster with seasonality-sensitive components
        EnsembleForecaster([
            ("ets", AutoETS(auto=True, n_jobs=1, additive_only=True, sp=sp)),
            ("arima", AutoARIMA(sp=sp))  # Seasonality often causes exceptions
        ]),
        # Second try: EnsembleForecaster without explicit seasonality
        EnsembleForecaster([
            ("ets", AutoETS(auto=True, n_jobs=1, additive_only=True, sp=None)),
            ("sarima", AutoARIMA(sp=None)),
        ]),
        # Final fallback: Simple ensemble with a naive forecaster
        EnsembleForecaster([
            ("ets", AutoETS(auto=True, n_jobs=1, additive_only=True, sp=None)),
            ("naive", NaiveForecaster(strategy="last")),
        ]),
    ]
)
```
There could be additional parameters for logging, performance tracking, customizable fallback logic, etc.

**Reply by fkiraly** [reply]
yes, exactly! There's some thinking for the parameters to do, e.g., when we consider an estimator as "failed" - a failure could occur in `fit`, `predict`, or `update`. And what that means for which estimators we fiti when.

**Reply by ninedigits** [reply]
I've got some starter code `FallbackForecaster` that works as intended when a forecaster fails at the `.fit()` step. Do you know of any situations where a forecaster will successfully fit but fail at `.predict()`?

EDIT: On second thought, I can just write a dummy forecaster that fails intentionally at the respective steps.

**Comment by fkiraly** [comment]
Another side note, @ninedigits, if you are already digging into `EnsembleForecaster`, it would be nice if you could replace the parallelization in `_HeterogenousEnsembleForecaster._fit_forecaster` by a call to the more general parallelization backend, see for example `ForecastingGridSearchCV` for how the arguments ought to look like.

This is a completely separate thinng though.

Relevant issue and recipe is here: https://github.com/sktime/sktime/issues/5380

**Reply by ninedigits** [reply]
Do you mean replace this parallel class:
```python3
from joblib import Parallel, delayed
...
self.forecasters_ = Parallel(n_jobs=self.n_jobs)(
    delayed(_fit_forecaster)(forecaster.clone(), y, X, fh)
    for forecaster in forecasters
)
```
with this?
```python3
from sktime.utils.parallel import parallelize
...
out = parallelize(
    fun=_fit_and_score,
    iter=candidate_params,
    backend=backend,
    backend_params=backend_params,
)
```

**Reply by fkiraly** [reply]
exactly! Of course you need to ensue that `backend` and `backend_params` are present in `__init__` and the docstring etc.
